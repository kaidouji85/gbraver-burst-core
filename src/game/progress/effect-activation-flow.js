// @flow

import type {GameState} from "../state/game-state";
import {gameFlow} from "./game-flow";
import {burst} from "../../effect/burst";
import {inputCommand} from "../../effect/input-command";
import type {PlayerCommand} from "../../command/command";
import {pilotSkill} from "../../effect/pilot-skill";

/**
 * 効果発動フローを行うか否かを判定する
 *
 * @param commands プレイヤーが選択したコマンド
 * @return 判定結果、trueでバーストフェイズを行う
 */
export function isEffectActivationFlow(commands: PlayerCommand[]): boolean {
  const types = commands.map(v => v.command.type);
  return types.includes('BURST_COMMAND') || types.includes('PILOT_SKILL_COMMAND');
}

/**
 * 効果発動フロー
 * 現状ではバースト、パイロットスキルを想定している
 *
 * @param lastState 最後の状態
 * @param commands コマンド
 * @return 更新されたゲームの状態
 */
export function effectActivationFlow(lastState: GameState, commands: PlayerCommand[]): GameState[] {
  const attackerCommand = commands.find(v => v.playerId === lastState.activePlayerId);
  const defenderCommand = commands.find(v => v.playerId !== lastState.activePlayerId);
  if (!attackerCommand || !defenderCommand) {
    return [];
  }

  return gameFlow(lastState, [
    state => activationOrNot(state, attackerCommand),
    state => activationOrNot(state, defenderCommand),
    state => [inputCommand(state, commands)],
  ]);
}

/**
 * コマンドに応じて 効果発動 or 何もしない
 *
 * @param state 最新の状態
 * @param command コマンド
 * @return 更新結果
 */
export function activationOrNot(state: GameState, command: PlayerCommand): GameState[] {
  switch(command.command.type) {
    case 'BURST_COMMAND':
      return [burst(state, command.playerId)];
    case 'PILOT_SKILL_COMMAND':
      return [pilotSkill(state, command.playerId)];
    default:
      return [];
  }
}