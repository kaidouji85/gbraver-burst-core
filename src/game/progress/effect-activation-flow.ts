import { burst } from "../../effect/burst";
import { inputCommand } from "../../effect/input-command";
import { pilotSkill } from "../../effect/pilot-skill";
import type { GameState } from "../../state/game-state";
import type { PlayerCommand } from "../command/player-command";
import { startGameStateFlow } from "../game-state-flow";

/**
 * 効果発動フローを行うか否かを判定する
 *
 * @param commands プレイヤーが選択したコマンド
 * @return 判定結果、trueでバーストフェイズを行う
 */
export function isEffectActivationFlow(
  commands: [PlayerCommand, PlayerCommand]
): boolean {
  const types = commands.map((v) => v.command.type);
  return (
    types.includes("BURST_COMMAND") || types.includes("PILOT_SKILL_COMMAND")
  );
}

/**
 * 効果発動フロー
 * 現状ではバースト、パイロットスキルを想定している
 *
 * @param lastState 最後の状態
 * @param commands コマンド
 * @return 更新されたゲームの状態
 */
export function effectActivationFlow(
  lastState: GameState,
  commands: [PlayerCommand, PlayerCommand]
): GameState[] {
  const attackerCommand = commands.find(
    (v) => v.playerId === lastState.activePlayerId
  );
  const defenderCommand = commands.find(
    (v) => v.playerId !== lastState.activePlayerId
  );

  if (!attackerCommand || !defenderCommand) {
    throw new Error("not found attacker or defender command");
  }

  return startGameStateFlow([lastState])
    .add((state) => {
      const done = activationOrNot(state, attackerCommand);
      return done ? [done] : [];
    })
    .add((state) => {
      const done = activationOrNot(state, defenderCommand);
      return done ? [done] : [];
    })
    .add((state) => {
      const done = inputCommand(
        state,
        attackerCommand.playerId,
        attackerCommand.command,
        defenderCommand.playerId,
        defenderCommand.command
      );
      return [done];
    })
    .toGameStateHistory()
    .slice(1); // 本関数は更新結果だけを返すので、ステートヒストリーの先頭は不要
}

/**
 * コマンドに応じて 効果発動 or 何もしない
 * 何もしない場合はnullを返す
 *
 * @param state 最新のゲームステート
 * @param command コマンド
 * @return 更新結果
 */
export function activationOrNot(
  state: GameState,
  command: PlayerCommand
): GameState | null {
  if (command.command.type === "BURST_COMMAND") {
    return burst(state, command.playerId);
  }

  if (command.command.type === "PILOT_SKILL_COMMAND") {
    return pilotSkill(state, command.playerId);
  }

  return null;
}
