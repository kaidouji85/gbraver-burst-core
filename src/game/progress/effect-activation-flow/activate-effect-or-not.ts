import { burst } from "../../../effect/burst";
import { pilotSkill } from "../../../effect/pilot-skill";
import { GameState } from "../../../state/game-state";
import { PlayerCommand } from "../../command/player-command";

/**
 * コマンドに応じて 効果発動 or 何もしない
 * 何もしない場合はnullを返す
 * @param state 最新のゲームステート
 * @param command コマンド
 * @returns 更新結果
 */
export function activateEffectOrNot(
  state: GameState,
  command: PlayerCommand,
): GameState | null {
  if (command.command.type === "BURST_COMMAND") {
    return burst(state, command.playerId);
  }

  if (command.command.type === "PILOT_SKILL_COMMAND") {
    return pilotSkill(state, command.playerId);
  }

  return null;
}
