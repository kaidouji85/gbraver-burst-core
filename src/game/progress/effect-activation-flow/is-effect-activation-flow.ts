import { PlayerCommand } from "../../command/player-command";

/**
 * 効果発動フローを行うか否かを判定する
 * @param commands プレイヤーが選択したコマンド
 * @returns 判定結果、trueでバーストフェイズを行う
 */
export function isEffectActivationFlow(
  commands: [PlayerCommand, PlayerCommand],
): boolean {
  const types = commands.map((v) => v.command.type);
  return (
    types.includes("BURST_COMMAND") || types.includes("PILOT_SKILL_COMMAND")
  );
}
