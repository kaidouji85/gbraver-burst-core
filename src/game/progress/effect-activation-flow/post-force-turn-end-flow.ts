import { inputCommand } from "../../../effect/input-command";
import { updateRemainingTurn } from "../../../effect/update-remaining-turn";
import { GameState } from "../../../state/game-state";
import { startGameFlow } from "../../game-flow";

/**
 * 強制ターンエンド後のフローを実行する
 * @param lastState 最新のゲームステート
 * @returns 更新されたゲームの状態
 */
export const postForceTurnEndFlow = (lastState: GameState): GameState[] =>
  startGameFlow(lastState, [
    (state) => [updateRemainingTurn(state)],
    (state) => [inputCommand({ lastState: state, noChoices: [] })],
  ]);
