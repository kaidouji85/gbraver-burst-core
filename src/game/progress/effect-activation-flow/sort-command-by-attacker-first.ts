import { PlayerId } from "../../../player/player";
import { PlayerCommand } from "../../command/player-command";

/**
 * 攻撃側が先頭になるようにコマンドをソートする
 * @param commands コマンド
 * @param lastState 最後の状態
 * @returns ソートされたコマンド
 */
export function sortCommandByAttackerFirst(
  commands: [PlayerCommand, PlayerCommand],
  lastState: { activePlayerId: PlayerId },
): [PlayerCommand, PlayerCommand] {
  const attackerCommand = commands.find(
    (v) => v.playerId === lastState.activePlayerId,
  );
  const defenderCommand = commands.find(
    (v) => v.playerId !== lastState.activePlayerId,
  );
  if (!attackerCommand || !defenderCommand) {
    throw new Error("not found attacker or defender command");
  }

  return [attackerCommand, defenderCommand];
}
