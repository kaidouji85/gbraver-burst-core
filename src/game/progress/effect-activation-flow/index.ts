import { inputCommand } from "../../../effect/input-command";
import { GameState } from "../../../state/game-state";
import { PlayerCommand } from "../../command/player-command";
import { startGameFlow } from "../../game-flow";
import { activateEffectOrNot } from "./activate-effect-or-not";

/**
 * 効果発動フロー
 * 現状ではバースト、パイロットスキルを想定している
 * @param lastState 最後の状態
 * @param commands コマンド
 * @returns 更新されたゲームの状態
 */
export function effectActivationFlow(
  lastState: GameState,
  commands: [PlayerCommand, PlayerCommand],
): GameState[] {
  const attackerCommand = commands.find(
    (v) => v.playerId === lastState.activePlayerId,
  );
  const defenderCommand = commands.find(
    (v) => v.playerId !== lastState.activePlayerId,
  );
  if (!attackerCommand || !defenderCommand) {
    throw new Error("not found attacker or defender command");
  }

  return startGameFlow(lastState, [
    (state) => {
      const done = activateEffectOrNot(state, attackerCommand);
      return done ? [done] : [];
    },
    (state) => {
      const done = activateEffectOrNot(state, defenderCommand);
      return done ? [done] : [];
    },
    (state) => [
      inputCommand({
        lastState: state,
        noChoices: commands.filter((c) => c.command.type === "BATTERY_COMMAND"),
      }),
    ],
  ]);
}
