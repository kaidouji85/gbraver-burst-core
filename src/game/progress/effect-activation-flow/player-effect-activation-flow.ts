import { updateRemainingTurn } from "../../../effect/update-remaining-turn";
import { GameState } from "../../../state/game-state";
import { PlayerCommand } from "../../command/player-command";
import { startGameFlow } from "../../game-flow";
import { activateEffectOrNot } from "./activate-effect-or-not";
import { inputCommand } from "../../../effect/input-command";

/** プレイヤー単体の効果発動フローの実行結果 */
type PlayerEffectActivationFlowResult = {
  /** 更新されたステート */
  stateHistory: GameState[];
  /** 次プレイヤーの効果発動をスキップするか、trueでスキップする */
  shouldNextEffectActivationSkip: boolean;
};

/**
 * プレイヤー単体の効果発動フローを実行する
 * @param lastState 最後のゲームステート
 * @param command プレイヤーが出したコマンド
 * @returns 実行結果
 */
export function playerEffectActivationFlow(
  lastState: GameState,
  command: PlayerCommand,
): PlayerEffectActivationFlowResult {
  const done = activateEffectOrNot(lastState, command);
  if (!done) {
    return { stateHistory: [], shouldNextEffectActivationSkip: false };
  }

  const isForceTurnEndActivated =
    done.effect.name === "BurstEffect" &&
    done.effect.burst.type === "ForceTurnEnd";
  if (isForceTurnEndActivated) {
    const postForceTurnEnd = startGameFlow(done, [
      (state) => [updateRemainingTurn(state)],
      (state) => [inputCommand({ lastState: state, noChoices: [] })],
    ]);
    return {
      stateHistory: [done, ...postForceTurnEnd],
      shouldNextEffectActivationSkip: true,
    };
  }

  return { stateHistory: [done], shouldNextEffectActivationSkip: false };
}
