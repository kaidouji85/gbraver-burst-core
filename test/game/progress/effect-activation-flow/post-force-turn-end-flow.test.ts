import {
  ArmdozerEffect,
  Burst,
  EMPTY_GAME_STATE,
  EMPTY_PLAYER_STATE,
  PilotSkill,
  PlayerState,
} from "../../../../src";
import { postForceTurnEndFlow } from "../../../../src/game/progress/effect-activation-flow/post-force-turn-end-flow";

/** バースト発動プレイヤーID */
const burstPlayerId = "burstPlayer";

/**
 * バースト発動プレイヤーを生成する
 * @param options 生成オプション
 * @param options.burst バースト
 * @param options.pilotSkill パイロットスキル
 * @param options.effects 適用中の効果
 * @returns バースト発動プレイヤー
 */
const createBurstPlayer = (options: {
  burst?: Burst;
  pilotSkill?: PilotSkill;
  effects?: ArmdozerEffect[];
}): PlayerState => ({
  ...EMPTY_PLAYER_STATE,
  playerId: burstPlayerId,
  armdozer: {
    ...EMPTY_PLAYER_STATE.armdozer,
    burst: options.burst ?? EMPTY_PLAYER_STATE.armdozer.burst,
    effects: options.effects ?? [],
    battery: 1,
    maxBattery: 5,
  },
  pilot: {
    ...EMPTY_PLAYER_STATE.pilot,
    skill: options.pilotSkill ?? EMPTY_PLAYER_STATE.pilot.skill,
  },
});

/** それ以外のプレイヤー */
const otherPlayer = {
  ...EMPTY_PLAYER_STATE,
  playerId: "otherPlayer",
};

test("強制ターンエンド発動後のフローが正しく実行される", () => {
  const burstPlayer = createBurstPlayer({
    burst: {
      type: "ForceTurnEnd",
      recoverBattery: 2,
    },
    effects: [
      {
        type: "HalveCorrectPower",
        period: {
          type: "TurnLimit",
          remainingTurn: 2,
        },
      },
      {
        type: "TryReflect",
        damage: 2000,
        effect: "Lightning",
        period: {
          type: "TurnLimit",
          remainingTurn: 1,
        },
      },
    ],
  });
  const lastState = {
    ...EMPTY_GAME_STATE,
    activePlayerId: otherPlayer.playerId,
    players: [otherPlayer, burstPlayer],
  };
  const result = postForceTurnEndFlow(lastState);
  expect(result).toMatchSnapshot();
});
