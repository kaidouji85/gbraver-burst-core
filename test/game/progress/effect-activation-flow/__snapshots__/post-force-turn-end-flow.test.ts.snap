// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`強制ターンエンド発動後のフローが正しく実行される 1`] = `
[
  {
    "activePlayerId": "otherPlayer",
    "effect": {
      "endArmdozerEffects": [
        {
          "effect": {
            "damage": 2000,
            "effect": "Lightning",
            "period": {
              "remainingTurn": 0,
              "type": "TurnLimit",
            },
            "type": "TryReflect",
          },
          "playerId": "burstPlayer",
        },
      ],
      "name": "UpdateRemainingTurn",
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "otherPlayer",
      },
      {
        "armdozer": {
          "battery": 1,
          "burst": {
            "recoverBattery": 2,
            "type": "ForceTurnEnd",
          },
          "effects": [
            {
              "period": {
                "remainingTurn": 1,
                "type": "TurnLimit",
              },
              "type": "HalveCorrectPower",
            },
          ],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "burstPlayer",
      },
    ],
  },
  {
    "activePlayerId": "otherPlayer",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "otherPlayer",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "burstPlayer",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "otherPlayer",
      },
      {
        "armdozer": {
          "battery": 1,
          "burst": {
            "recoverBattery": 2,
            "type": "ForceTurnEnd",
          },
          "effects": [
            {
              "period": {
                "remainingTurn": 1,
                "type": "TurnLimit",
              },
              "type": "HalveCorrectPower",
            },
          ],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "burstPlayer",
      },
    ],
  },
]
`;
